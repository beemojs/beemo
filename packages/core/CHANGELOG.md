# 1.0.0

#### üí• Breaking

- Migrated to the new [@boost/event](https://milesj.gitbook.io/boost/event) system.
- Script names (on the command line) will now error if not in kebab case.
- Scripts and Drivers now require an explicit `blueprint` method.
- Renamed `--live` option to `--stdio`.
- Renamed `--priority` option to `--graph`.
- Renamed `execute.priority` setting to `execute.graph`.
- Renamed `Context#root` to `cwd`.
- Renamed `Driver#handleFailure` to `processFailure`.
- Renamed `Driver#handleSuccess` to `processSuccess`.
- Removed `DriverContext#eventName` and `ScriptContext#eventName`.
- Removed `ScriptContext#binName`. Use `scriptName` instead.

#### üöÄ Updates

- Added a new package,
  [@beemo/dependency-graph](https://www.npmjs.com/package/@beemo/dependency-graph), to handle the
  dependency resolution.
- Added a `none` strategy to `Driver`s. With this strategy, the consumer will need to manually
  create a config file.
- Added a `versionOption` metadata setting to `Driver`s.
- Added `Script#executeCommand`, so that local binaries can easily be executed.
- Added optional `name` support to scaffolding.
- Updated the driver/script execution pipeline to process in parallel batches.

#### üêû Fixes

- Fixed script `this` scope being lost within script tasks.
- Fixed the Beemo emoji not appearing in the console.

#### üõ† Internals

- `Beemo` now extends from `Tool` instead of managing an instance.
- Updated `hygen` to v4.
- Updated dependencies.
