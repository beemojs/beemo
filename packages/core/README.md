# ðŸ¤– Beemo

[![Build Status](https://github.com/beemojs/beemo/workflows/Build/badge.svg)](https://github.com/beemojs/beemo/actions?query=branch%3Amaster)
[![npm version](https://badge.fury.io/js/%40beemo%2Fcore.svg)](https://www.npmjs.com/package/@beemo/core)
[![npm deps](https://david-dm.org/beemojs/beemo.svg?path=packages/core)](https://www.npmjs.com/package/@beemo/core)

Manage dev and build tools, their configuration, and commands in a single centralized repository.
Beemo aims to solve the multi-project maintenance fatigue by removing the following burdens across
all projects: config and dotfile management, multiple config patterns, up-to-date development
dependencies, continuous copy and paste, and more.

### Features

- Manage dev tools and configurations in a single repository.
- Configure supported dev tools using `.js` files.
- Customize and alter config at runtime with CLI options.
- Pass custom CLI options to dev tool commands without failure.
- Automatically expand glob patterns (a better alternative to bash).
- Listen to and act upon events.
- Easily share config between dev tools.
- Avoid relative config or `extends` paths.
- Automatic config file cleanup.
- Custom scripts with CLI options.
- Scaffolding and template generation.
- Workspaces (monorepo) support.
- Automated TypeScript project references.
- Parallel, pooled, and prioritized builds.
- Integrated watch/live mode.
- And much more.

## Requirements

- Node 8.9+
- GitHub, Bitbucket, or another VCS

## Documentation

[https://milesj.gitbook.io/beemo](https://milesj.gitbook.io/beemo)
