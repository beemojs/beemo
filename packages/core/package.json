{
  "name": "@beemo/core",
  "version": "2.0.0-alpha.0",
  "description": "Manage all dev tool configurations in a single centralized repository.",
  "keywords": [
    "beemo",
    "build",
    "tool",
    "config",
    "manager",
    "babel",
    "eslint",
    "flow",
    "jest",
    "prettier",
    "typescript"
  ],
  "main": "./lib/index.js",
  "types": "./lib/index.d.ts",
  "engines": {
    "node": ">=10.17.0"
  },
  "repository": "https://github.com/beemojs/beemo/tree/master/packages/core",
  "author": "Miles Johnson",
  "license": "MIT",
  "publishConfig": {
    "access": "public"
  },
  "dependencies": {
    "@beemo/dependency-graph": "^2.0.0-alpha.0",
    "@boost/args": "^2.3.1",
    "@boost/common": "^2.5.1",
    "@boost/config": "^2.3.1",
    "@boost/debug": "^2.2.1",
    "@boost/event": "^2.3.1",
    "@boost/pipeline": "^2.2.1",
    "@boost/plugin": "^2.2.1",
    "@boost/terminal": "^2.2.0",
    "@boost/translate": "^2.2.1",
    "chalk": "^4.1.0",
    "execa": "^5.0.0",
    "fast-glob": "^3.2.5",
    "fs-extra": "^9.1.0",
    "hygen": "^6.0.4",
    "is-glob": "^4.0.1",
    "lodash": "^4.17.15",
    "micromatch": "^4.0.2"
  },
  "devDependencies": {
    "@boost/test-utils": "^2.2.0",
    "@types/micromatch": "^4.0.1"
  }
}
