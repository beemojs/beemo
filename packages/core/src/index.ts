/**
 * @copyright   2021, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import { Blueprint, PackageStructure, Path, PortablePath, Predicates } from '@boost/common';

export * from './constants';
export * from './contexts/ConfigContext';
export * from './contexts/Context';
export * from './contexts/DriverContext';
export * from './contexts/ScaffoldContext';
export * from './contexts/ScriptContext';
export * from './Driver';
export * from './Script';
export * from './Tool';
export * from './types';

export { Path };
export type { Blueprint, PackageStructure, PortablePath, Predicates };
