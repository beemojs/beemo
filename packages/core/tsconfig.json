{
  "compilerOptions": {
    "declarationDir": "lib",
    "outDir": "lib",
    "rootDir": "src"
  },
  "exclude": [
    "lib",
    "tests"
  ],
  "extends": "../../tsconfig.options.json",
  "include": [
    "src/**/*",
    "types/**/*",
    "../../types/**/*"
  ],
  "references": []
}