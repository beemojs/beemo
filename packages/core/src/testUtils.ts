/* eslint-disable max-classes-per-file */

import path from 'path';
import execa from 'execa';
import parseArgs from 'yargs-parser';
import { mockDebugger, stubArgs, stubToolConfig } from '@boost/core/test-utils';
import Beemo from './Beemo';
import Driver from './Driver';
import Script from './Script';
import Context from './contexts/Context';
import ConfigContext, { ConfigArgs } from './contexts/ConfigContext';
import DriverContext, { DriverArgs } from './contexts/DriverContext';
import ScaffoldContext, { ScaffoldArgs } from './contexts/ScaffoldContext';
import ScriptContext, { ScriptArgs } from './contexts/ScriptContext';
import { DriverMetadata, Argv, BeemoConfig } from './types';

export { mockDebugger, stubArgs };

export class TestDriver<T extends object = {}> extends Driver<T> {
  name = 'test-driver';
}

export class TestScript<A extends object = {}, T extends object = {}> extends Script<A, T> {
  name = 'test-script';

  blueprint() {
    return {} as any;
  }
}

// Use core package since resources are located here
export const BEEMO_APP_PATH = path.join(__dirname, '..');

// Use a folder that should not cause issues / contain much code
export const BEEMO_TEST_ROOT = path.join(__dirname, '../../../tests');

export function mockTool(argv: Argv = []): Beemo {
  const tool = new Beemo(argv, '', true);

  Object.assign(tool.options, {
    appName: 'beemo',
    appPath: BEEMO_APP_PATH,
    root: BEEMO_TEST_ROOT,
    workspaceRoot: BEEMO_TEST_ROOT,
  });

  tool.config = stubToolConfig<BeemoConfig>({
    configure: {
      cleanup: false,
      parallel: true,
    },
    drivers: [],
    execute: {
      concurrency: 0,
      priority: true,
    },
    scripts: [],
  });

  return tool;
}

export function mockDriver<C extends object = {}>(
  name: string,
  tool: Beemo | null = null,
  metadata: Partial<DriverMetadata> = {},
): Driver<C> {
  const driver = new TestDriver<C>();

  driver.name = name;
  driver.tool = tool || mockTool();

  driver.setMetadata({
    bin: name,
    configName: `${name}.json`,
    title: name,
    ...metadata,
  });

  driver.bootstrap();

  return driver;
}

export function mockScript<C extends object = {}>(
  name: string,
  tool: Beemo | null = null,
): Script<{}, C> {
  const script = new TestScript<{}, C>();

  script.name = name;
  script.tool = tool || mockTool();

  return script;
}

export function applyContext<T extends Context>(context: T): T {
  context.args = parseArgs(['-a', '--foo', 'bar', 'baz']);
  context.argv = ['-a', '--foo', 'bar', 'baz'];
  context.cwd = BEEMO_TEST_ROOT;
  context.moduleRoot = BEEMO_TEST_ROOT;
  context.workspaceRoot = BEEMO_TEST_ROOT;
  context.workspaces = [];

  return context;
}

export function stubContext(): Context {
  return applyContext(new Context(stubArgs()));
}

export function stubConfigArgs(fields?: Partial<ConfigArgs>) {
  return stubArgs<ConfigArgs>({
    names: [],
    ...fields,
  });
}

export function stubConfigContext(): ConfigContext {
  return applyContext(new ConfigContext(stubConfigArgs()));
}

export function stubDriverArgs(fields?: Partial<DriverArgs>) {
  return stubArgs<DriverArgs>({
    concurrency: 1,
    live: false,
    priority: false,
    workspaces: '',
    ...fields,
  });
}

export function stubDriverContext(driver?: Driver): DriverContext {
  return applyContext(new DriverContext(stubDriverArgs(), driver || new TestDriver()));
}

export function stubScaffoldArgs(fields?: Partial<ScaffoldArgs>) {
  return stubArgs<ScaffoldArgs>({
    action: '',
    dry: false,
    generator: '',
    name: '',
    ...fields,
  });
}

export function stubScaffoldContext(
  generator: string = 'generator',
  action: string = 'action',
  name: string = '',
): ScaffoldContext {
  return applyContext(new ScaffoldContext(stubScaffoldArgs(), generator, action, name));
}

export function stubScriptArgs(fields?: Partial<ScriptArgs>) {
  return stubArgs<ScriptArgs>({
    concurrency: 1,
    name: 'foo',
    priority: false,
    workspaces: '',
    ...fields,
  });
}

export function stubScriptContext(script?: Script): ScriptContext {
  const context = applyContext(new ScriptContext(stubScriptArgs(), 'script'));

  if (script) {
    context.setScript(script, './script.js');
  }

  return context;
}

export function stubExecResult(fields?: Partial<execa.ExecaReturns>): execa.ExecaReturns {
  return {
    cmd: '',
    code: 0,
    failed: false,
    killed: false,
    signal: null,
    stderr: '',
    stdout: '',
    timedOut: false,
    ...fields,
  };
}

export function prependRoot(part: string): string {
  return path.join(BEEMO_TEST_ROOT, part);
}

export function getRoot(): string {
  return BEEMO_TEST_ROOT;
}
