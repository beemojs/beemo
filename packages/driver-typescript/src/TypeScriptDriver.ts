/* eslint-disable no-param-reassign */

import rimraf from 'rimraf';
import { Blueprint, ConfigContext, Driver, DriverContext, Path, Schemas } from '@beemo/core';
import { Event } from '@boost/event';
import { syncProjectRefs } from './commands/syncProjectRefs';
import { toForwardSlashes, writeFile } from './helpers';
import { TypeScriptConfig, TypeScriptOptions } from './types';

// Success: Writes nothing to stdout or stderr
// Failure: Writes to stdout on syntax and type error
export class TypeScriptDriver extends Driver<TypeScriptConfig, TypeScriptOptions> {
	override readonly name = '@beemo/driver-typescript';

	readonly onCreateProjectConfigFile = new Event<[Path, TypeScriptConfig, boolean]>(
		'create-project-config-file',
	);

	override blueprint(schemas: Schemas): Blueprint<TypeScriptOptions> {
		const { bool, string } = schemas;

		return {
			...super.blueprint(schemas),
			buildFolder: string('lib'),
			declarationOnly: bool(),
			globalTypes: bool(true),
			localTypes: bool(true),
			srcFolder: string('src'),
			testsFolder: string('tests'),
			typesFolder: string('types'),
		};
	}

	override bootstrap() {
		this.setMetadata({
			bin: 'tsc',
			commandOptions: {
				clean: {
					default: false,
					description: this.tool.msg('app:typescriptCleanOption'),
					type: 'boolean',
				},
			},
			configName: 'tsconfig.json',
			configOption: '',
			description: this.tool.msg('app:typescriptDescription'),
			helpOption: '--help --all',
			title: 'TypeScript',
			watchOptions: ['-w', '--watch'],
			workspaceStrategy: 'copy',
		});

		this.registerCommand(
			'sync-project-refs',
			{ description: this.tool.msg('app:typescriptSyncProjectRefsDescription') },
			syncProjectRefs,
		);

		this.onCreateConfigFile.listen(this.handlePrepareConfigs);
		this.onBeforeExecute.listen(this.handleCleanTarget);
	}

	/**
	 * Automatically clean the target folder if `outDir` and `--clean` is used.
	 */
	handleCleanTarget = (context: DriverContext) => {
		// eslint-disable-next-line @typescript-eslint/prefer-nullish-coalescing
		const outDir = context.getRiskyOption('outDir', true) || this.config.compilerOptions?.outDir;

		if (context.getRiskyOption('clean') && typeof outDir === 'string' && outDir) {
			rimraf.sync(Path.resolve(outDir).path());
		}

		return Promise.resolve();
	};

	/**
	 * Extract compiler options from the root config into a separate config purely for
	 * extending options. Update the root config with references to all workspaces.
	 */
	handlePrepareConfigs = (context: ConfigContext, configPath: Path, config: TypeScriptConfig) => {
		const { tool } = this;
		const { srcFolder, testsFolder } = this.options;
		const workspacePackages = tool.project.getWorkspacePackages();

		if (workspacePackages.length === 0) {
			return;
		}

		// Extract compiler optionst to a separate config
		const optionsConfigPath = configPath.parent().append('tsconfig.options.json');

		void writeFile(optionsConfigPath, {
			compilerOptions: {
				...config.compilerOptions,
				composite: true,
				declaration: true,
				declarationMap: true,
				outDir: undefined,
				outFile: undefined,
			},
		});

		// Delete problematic root options
		delete config.compilerOptions;
		delete config.include;
		delete config.exclude;

		// Generate references and update paths
		config.extends = './tsconfig.options.json';
		config.files = [];
		config.references ||= [];

		workspacePackages.forEach(({ metadata }) => {
			const pkgPath = new Path(metadata.packagePath);
			const srcPath = pkgPath.append(srcFolder);
			const testsPath = pkgPath.append(testsFolder);

			// Reference a package *only* if it has a src folder
			if (srcFolder && srcPath.exists()) {
				config.references!.push({
					path: toForwardSlashes(tool.project.root.relativeTo(pkgPath).path()),
				});

				// Reference a separate tests folder if it exists
				if (testsFolder && testsPath.exists()) {
					config.references!.push({
						path: toForwardSlashes(tool.project.root.relativeTo(testsPath).path()),
					});
				}
			}
		});

		// Add to context so that it can be automatically cleaned up
		context.addConfigPath('typescript', optionsConfigPath);
	};
}
