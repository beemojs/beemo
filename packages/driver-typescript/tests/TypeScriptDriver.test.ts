import path from 'path';
import rimraf from 'rimraf';
import { DriverContext } from '@beemo/core';
import { mockTool, stubDriverContext } from '@beemo/core/lib/testUtils';
import TypeScriptDriver from '../src/TypeScriptDriver';

jest.mock('rimraf');

describe('TypeScriptDriver', () => {
  let driver: TypeScriptDriver;
  let context: DriverContext;

  beforeEach(() => {
    driver = new TypeScriptDriver();
    driver.tool = mockTool();
    driver.bootstrap();
    driver.config = {
      compilerOptions: {},
    };

    context = stubDriverContext(driver);
  });

  it('sets options from constructor', () => {
    driver = new TypeScriptDriver({
      args: ['--foo', '--bar=1'],
      dependencies: ['babel'],
      env: { DEV: 'true' },
    });

    expect(driver.options).toEqual({
      args: ['--foo', '--bar=1'],
      dependencies: ['babel'],
      env: { DEV: 'true' },
      strategy: 'native',
      buildFolder: 'lib',
      srcFolder: 'src',
      testsFolder: 'tests',
    });
  });

  it('sets correct metadata', () => {
    expect(driver.metadata).toEqual(
      expect.objectContaining({
        bin: 'tsc',
        configName: 'tsconfig.json',
        configOption: '',
        dependencies: [],
        description: 'Type check files with TypeScript',
        filterOptions: true,
        helpOption: '--help --all',
        title: 'TypeScript',
        useConfigOption: false,
        workspaceStrategy: 'copy',
      }),
    );
  });

  describe('handleCleanTarget()', () => {
    it('doesnt run if no config', () => {
      driver.config = {};
      driver.handleCleanTarget(context);

      expect(rimraf.sync).not.toHaveBeenCalled();
    });

    it('doesnt run if no clean param', () => {
      driver.config.compilerOptions = { outDir: './lib' };
      driver.handleCleanTarget(context);

      expect(rimraf.sync).not.toHaveBeenCalled();
    });

    it('doesnt run if no outDir param', () => {
      context.args.clean = true;
      driver.config.compilerOptions = {};
      driver.handleCleanTarget(context);

      expect(rimraf.sync).not.toHaveBeenCalled();
    });

    it('runs if both params', () => {
      context.args.clean = true;
      driver.config.compilerOptions = { outDir: './lib' };
      driver.handleCleanTarget(context);

      expect(rimraf.sync).toHaveBeenCalledWith(path.resolve('./lib'));
    });
  });
});
