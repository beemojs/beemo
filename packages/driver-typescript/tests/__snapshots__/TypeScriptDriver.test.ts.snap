// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TypeScriptDriver handleProjectReferences() errors if --reference-workspaces and --workspaces are passed 1`] = `"Options --workspaces and --reference-workspaces cannot be used in unison."`;
